File: start-code/symtab.c
/**********************************************************************/
/* lab 1 DVG C01 - Symbol Table OBJECT                                */
/**********************************************************************/

/**********************************************************************/
/* Include files                                                      */
/**********************************************************************/
#include <stdio.h>
#include <ctype.h>
#include <string.h>

/**********************************************************************/
/* Other OBJECT's METHODS (IMPORTED)                                  */
/**********************************************************************/
#include "keytoktab.h"

/**********************************************************************/
/* OBJECT ATTRIBUTES FOR THIS OBJECT (C MODULE)                       */
/**********************************************************************/
#define TABSIZE 1024                   /* symbol table size           */
#define NAMELEN   20                   /* name length                 */

typedef char tname[NAMELEN];

static tname  name[TABSIZE];
static toktyp role[TABSIZE];
static toktyp type[TABSIZE];
static int    size[TABSIZE];
static int    addr[TABSIZE];

static int numrows=0;                  /* number of rows in the ST    */
static int startp =0;                  /* start position program in ST*/

/**********************************************************************/
/*  PRIVATE METHODS for this OBJECT  (using "static" in C)            */
/**********************************************************************/
/**********************************************************************/
/*  GET methods (one for each attribute)                              */
/**********************************************************************/
static char * get_name(int ftref)   { return name[ftref]; }
static toktyp get_role(int ftref)   { return role[ftref]; }
static toktyp get_type(int ftref)   { return type[ftref]; }
static int    get_size(int ftref)   { return size[ftref]; }
static int    get_addr(int ftref)   { return addr[ftref]; }

/**********************************************************************/
/*  SET methods (one for each attribute)                              */
/**********************************************************************/
static void set_name(int ftref, char * fpname) { strcpy(name[ftref],
                                                        fpname);}
static void set_role(int ftref, toktyp frole)  { role[ftref] = frole; }
static void set_type(int ftref, toktyp ftype)  { type[ftref] = ftype; }
static void set_size(int ftref, int    fsize)  { size[ftref] = fsize; }
static void set_addr(int ftref, int    faddr)  { addr[ftref] = faddr; }

/**********************************************************************/
/*  Add a row to the symbol table                                     */
/**********************************************************************/
static void addrow(char *fname, toktyp frole, toktyp ftype,
                   int fsize, int faddr)
{
    set_name(numrows, fname);
    set_role(numrows, frole);
    set_type(numrows, ftype);
    set_size(numrows, fsize);
    set_addr(numrows++, faddr);
}
/**********************************************************************/
/*  Initialise the symbol table                                       */
/**********************************************************************/
static void initst()
{
    addrow(tok2lex(predef),  typ, predef, 0, 0);
    addrow(tok2lex(undef),   typ, predef, 0, 0);
    addrow(tok2lex(error),   typ, predef, 0, 0);
    addrow(tok2lex(integer), typ, predef, 4, 0);
    addrow(tok2lex(boolean), typ, predef, 4, 0);
    addrow(tok2lex(real),    typ, predef, 8, 0);
}
/**********************************************************************/
/*  return a reference to the ST (index) if name found else nfound    */
/**********************************************************************/
static int get_ref(char * fpname)
{
    printf("\n *** TO BE DONE"); return 0;
}

/**********************************************************************/
/*  PUBLIC METHODS for this OBJECT  (EXPORTED)                        */
/**********************************************************************/
/**********************************************************************/
/*  Display the symbol table                                          */
/**********************************************************************/
static void p_symrow(int ftref)
{
    printf("\n *** TO BE DONE");
}

void p_symtab()
{
    initst();
    printf("________________________________________________________\n");
    printf(" THE SYMBOL TABLE\n");
    printf("________________________________________________________\n");
    printf("%11s%11s%11s%10s%10s\n", "NAME", "ROLE", "TYPE", "SIZE", "ADDR");
    printf("________________________________________________________\n");
    for (int i = 0; i < numrows; i++) {
        printf("%11s%11s%11s%10d%10d\n", name[i], tok2lex(role[i]), tok2lex(type[i]), size[i], addr[i]);
    }
    printf("________________________________________________________\n");
    printf("________________________________________________________\n");
    //  STATIC STORAGE REQUIRED is 0 BYTES
    printf("________________________________________________________\n");
}

/**********************************************************************/
/*  Add a program name to the symbol table                            */
/**********************************************************************/
void addp_name(char * fpname)
{
    strcpy(name[0],fpname);
    set_role(numrows, program);
}

/**********************************************************************/
/*  Add a variable name to the symbol table                           */
/**********************************************************************/
void addv_name(char * fpname)
{
    strcpy(name[++numrows],fpname);
    set_role(numrows, var);
}

/**********************************************************************/
/*  Find a name in the the symbol table                               */
/*  return a Boolean (true, false) if the name is in the ST           */
/**********************************************************************/
int find_name(char * fpname)
{
    printf("\n *** TO BE DONE"); return 0;
}

/**********************************************************************/
/*  Set the type of an id list in the symbol table                    */
/**********************************************************************/
void setv_type(toktyp ftype)
{
    printf("\n *** TO BE DONE");
}

/**********************************************************************/
/*  Get the type of a variable from the symbol table                  */
/**********************************************************************/
toktyp get_ntype(char * fpname)
{
    printf("\n *** TO BE DONE"); return 0;
}

/**********************************************************************/
/* End of code                                                        */
/**********************************************************************/

File: Lab 4 - 🐍/symtab.py
from tokens import TokenType, tok2lex


class SymEntry:
    def __init__(self, name, role, typ, size, addr):
        self.name = name
        self.role = role
        self.type = typ
        self.size = size
        self.addr = addr


class SymTab:
    def __init__(self):
        self.entries = []
        self.program_size = 0

    def addp_name(self, name):
        self.entries.append(SymEntry(
            name,
            TokenType.program,
            TokenType.program,
            0,
            0))

    def addv_name(self, name):
        self.entries.append(SymEntry(name, TokenType.var, None, 0, 0))

    def find_name(self, name):
        return any(e.name == name for e in self.entries)

    def setv_type(self, typ):
        for e in reversed(self.entries):
            if e.type is not None:
                break
            e.type = typ
            if typ in (TokenType.integer, TokenType.boolean):
                e.size = 4
            elif typ == TokenType.real:
                e.size = 8
            else:
                e.size = 0
            self.program_size += e.size
        addr = 0
        for e in self.entries[1:]:
            e.addr = addr
            addr += e.size
        self.program_size = addr
        if self.entries:
            self.entries[0].size = self.program_size

    def get_ntype(self, name):
        for e in self.entries:
            if e.name == name:
                return e.type if e.type is not None else TokenType.nfound
        return TokenType.nfound

    def p_symtab(self):
        print("________________________________________________________")
        print(" THE SYMBOL TABLE")
        print("________________________________________________________")
        print(f"{'NAME':>11}{'ROLE':>11}{'TYPE':>11}{'SIZE':>10}{'ADDR':>10}")
        print("________________________________________________________")
        for e in self.entries:
            tlex = tok2lex(e.type) if e.type is not None else 'undef'
            print(f"{e.name:>11}{tok2lex(e.role):>11}{tlex:>11}{e.size:10d}{e.addr:10d}")
        print("________________________________________________________")
        print(f" STATIC STORAGE REQUIRED is {self.program_size} BYTES")
        print("________________________________________________________")